### PR TITLE
Reduce linear search on list traversing

### LINKED ISSUE #11527: traverse of Array Literal takes O(n^2) time
## Bug Report

- [ ] I would like to work on a fix!

**Current Behavior**

Babel's traverse module takes O(n^2) time to traverse an Array Literal, which causes some unexpected slow-downs. I suspect this is due to consistent worst-case performance on the path cache [in `babel-traverse/src/path/index.js:89`](https://github.com/babel/babel/blob/f6c7bf36cec81baaba8c37e572985bb59ca334b1/packages/babel-traverse/src/path/index.js#L89). During traversal, each of the `n` members of an array literal will undergo path construction. As part of this, the parent node (the array literal) will be consulted for a cache (list) of known child paths. A linear scan of order `n` is performed on this cache, which seems to hit the worst-case (cache miss) each time. Hence, O(n^2) performance.

**Input Code**

See: https://gist.github.com/robmcl4/080bf388c95de3b2761307b39d6f895a

**Expected behavior/code**

I expected the traversal to show execution time growing roughly linearly with array size.

**Babel Configuration (babel.config.js, .babelrc, package.json#babel, cli command, .eslintrc)**
- Filename: `babel.config.js`

**Environment**

```
  System:
    OS: Linux 5.3 Ubuntu 18.04.3 LTS (Bionic Beaver)
  Binaries:
    Node: 12.16.2 - /usr/bin/node
    npm: 6.14.5 - /usr/bin/npm
  npmPackages:
    @babel/parser: ^7.9.6 => 7.9.6
    @babel/traverse: ^7.9.6 => 7.9.6
    eslint: ^6.8.0 => 6.8.0
```

- How you are using Babel: research project involving traversal of the AST of popular npm packages

**Possible Solution**

Potentially remove the path cache for traversal, or modify it to use a Set instead of a list, or carve out some sort of special case for array literals.

**Additional context/Screenshots**

Included below is a plot of the output from an extended version (more elems) of `what.js`, linked above. The thin line demonstrates a quadratic function fit.

image

### PR DESCRIPTION
**Pull request by JLHwung**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Fixes #11527 
| Tests Added + Pass?      | No, see below
| Documentation PR Link    | 
| Any Dependency Changes?  |
| License                  | MIT

Checkout this branch, build Babel and run https://gist.github.com/robmcl4/080bf388c95de3b2761307b39d6f895a

The output on my local machine is attached here for reference
<details>
  <summary>Current main</summary>

```
2,0.049
4,0.037
8,0.017
16,0.018
32,0.028
64,0.057
128,0.114
256,0.306
512,0.900
1024,3.106
2048,11.738
4096,45.160
```
</details>

<details>
  <summary>This PR</summary>

```
2,0.046
4,0.043
8,0.022
16,0.020
32,0.028
64,0.053
128,0.083
256,0.164
512,0.336
1024,0.658
2048,1.343
4096,2.764
```
</details>

In other words, traversing an ArrayLiteral of 4096 elements can be up to 16x faster. The execution time also looks linear.

Note that this issue is not limited to ArrayLiteral, I can reproduce it on any list-style AST structures: e.g. Program, ObjectLiteral, BlockStatement, etc. Although it is uncommon that an array literal may contain 4096 elements, a Program/BlockStatement, especially bundled, could contain thousands of statements as their list children.

<a href="https://gitpod.io/#https://github.com/babel/babel/pull/12302"></a>

### CODE REVIEW

**Review by kaicataldo — APPROVED** [review]
🎉

**Comment by nicolo-ribaudo** [comment]
This is awesome :smile:

**Comment by JLHwung** [comment]
@nicolo-ribaudo @kaicataldo Thanks! I converted this PR to draft since it is breaking TDZ implementation, which I am investigating.

File: packages/babel-traverse/src/context.js [hunk 1]
```diff
@@ -120,8 +120,9 @@ export default class TraversalContext {
       }
 
       // ensure we don't visit the same node twice
-      if (visited.indexOf(path.node) >= 0) continue;
-      visited.push(path.node);
+      const { node } = path;
+      if (visited.has(node)) continue;
+      if (node) visited.add(node);
```
Comment (JLHwung) [inline packages/babel-traverse/src/context.js:125]: Note that when `path.node` is `null`, the current behaviour is to visit this path once but skip other nullish `node` paths since `null` can be hold in array. In this PR they will all be visited. This could be a breaking change in a way but if two paths whose `path.node` is null, I think their equality is not well-defined.

The usage of `WeakSet` is intentional, so we don't leak memory when `path.node` is GCed.

File: packages/babel-traverse/src/path/index.js [hunk 2]
```diff
@@ -67,24 +67,16 @@ export default class NodePath {
 
     const targetNode = container[key];
 
-    const paths = pathCache.get(parent) || [];
-    if (!pathCache.has(parent)) {
+    let paths = pathCache.get(parent);
+    if (!paths) {
+      paths = new Map();
```
Comment (JLHwung) [inline packages/babel-traverse/src/path/index.js:72]: We can't use a `WeakMap` here because we iterate on pathCache in https://github.com/babel/babel/pull/12302/files#diff-e406da86b11f8d4846c6ffb3bf527e1c0fa230888127f90bb8978c16a9bd025aR166

Comment (nicolo-ribaudo) [reply packages/babel-traverse/src/path/index.js:72]: We can create an `IterableWeakMap`, falling back to a normal `Map` on Node.js <13

https://github.com/nodejs/node/pull/35915

Comment (JLHwung) [reply packages/babel-traverse/src/path/index.js:72]: Oh the `IterableWeakMap` is not a builtin object nor supposed to be a public Node.js API, I suggest we hold off a bit since `WeakRef` (relied by `IterableWeakMap`) is still stage 3 and `@babel/traverse` is heavily used.

**Review by jridgewell — APPROVED** [review]

**Review by nicolo-ribaudo — APPROVED** [review]

**Comment by jedwards1211** [comment]
@JLHwung I just bisected and determined that this introduced the regression #12570, do you have any thoughts on that?
